src/encoding.rs

```diff
@@ -0,0 +1,197 @@
+        self.encoding
+    }
+
+    /// Decodes specified bytes using encoding, declared in the XML, if it was
+    /// declared there, or UTF-8 otherwise, and ignoring BOM if it is present
+    /// in the `bytes`.
+    ///
+    /// Returns an error in case of malformed sequences in the `bytes`.
+    pub fn decode<'b>(&self, bytes: &'b [u8]) -> Result<Cow<'b, str>> {
+        decode(bytes, self.encoding)
+    }
+
+    /// Decodes a slice with BOM removal if it is present in the `bytes` using
+    /// the reader encoding.
+    ///
+    /// If this method called after reading XML declaration with the `"encoding"`
+    /// key, then this encoding is used, otherwise UTF-8 is used.
+    ///
+    /// If XML declaration is absent in the XML, UTF-8 is used.
+    ///
+    /// Returns an error in case of malformed sequences in the `bytes`.
+    pub fn decode_with_bom_removal<'b>(&self, bytes: &'b [u8]) -> Result<Cow<'b, str>> {
+        self.decode(remove_bom(bytes, self.encoding))
+    }
+}
+
+/// Decodes the provided bytes using the specified encoding, ignoring the BOM
+/// if it is present in the `bytes`.
+///
+/// Returns an error in case of malformed sequences in the `bytes`.
+#[cfg(feature = "encoding")]
+pub fn decode<'b>(bytes: &'b [u8], encoding: &'static Encoding) -> Result<Cow<'b, str>> {
+    encoding
+        .decode_without_bom_handling_and_without_replacement(bytes)
+        .ok_or(Error::NonDecodable(None))
+}
+
+/// Decodes a slice with an unknown encoding, removing the BOM if it is present
+/// in the bytes.
+///
+/// Returns an error in case of malformed sequences in the `bytes`.
+#[cfg(feature = "encoding")]
+pub fn decode_with_bom_removal<'b>(bytes: &'b [u8]) -> Result<Cow<'b, str>> {
+    if let Some(encoding) = detect_encoding(bytes) {
+        let bytes = remove_bom(bytes, encoding);
+        decode(bytes, encoding)
+    } else {
+        decode(bytes, UTF_8)
+    }
+}
+
+#[cfg(feature = "encoding")]
+fn split_at_bom<'b>(bytes: &'b [u8], encoding: &'static Encoding) -> (&'b [u8], &'b [u8]) {
+    if encoding == UTF_8 && bytes.starts_with(&[0xEF, 0xBB, 0xBF]) {
+        bytes.split_at(3)
+    } else if encoding == UTF_16LE && bytes.starts_with(&[0xFF, 0xFE]) {
+        bytes.split_at(2)
+    } else if encoding == UTF_16BE && bytes.starts_with(&[0xFE, 0xFF]) {
+        bytes.split_at(2)
+    } else {
+        (&[], bytes)
+    }
+}
+
+#[cfg(feature = "encoding")]
+fn remove_bom<'b>(bytes: &'b [u8], encoding: &'static Encoding) -> &'b [u8] {
+    let (_, bytes) = split_at_bom(bytes, encoding);
+    bytes
+}
+
+/// Automatic encoding detection of XML files based using the [recommended algorithm]
+/// (https://www.w3.org/TR/xml11/#sec-guessing)
+///
+/// The algorithm suggests examine up to the first 4 bytes to determine encoding
+/// according to the following table:
+///
+/// | Bytes       |Detected encoding
+/// |-------------|------------------------------------------
+/// |`00 00 FE FF`|UCS-4, big-endian machine (1234 order)
```

Comment:
Because now the function is public, maybe more explicitly mark rows, where autodetection is supported?

Reply:
I've just removed them entirely.  It is unlikely that support for those encodings will ever be available, at least using this library.